File: codes/array/704.BinarySearch/BinarySearch.cpp
#include <iostream>
#include <vector>
#include <fmt/core.h>
using namespace std;
// 2023年07月06日 修改为 acwing版本
// 找右边界
// class Solution {
// public:
//     int search(vector<int>& nums, int target) {
//         int mid, l = 0, r = nums.size() - 1;
//         while (l < r) {
//             mid = l + r + 1 >> 1;
//             if (nums[mid] <= target) {
//                 l = mid;
//             } else {
//                 r = mid - 1;
//             }
//         }
//         if (nums[l] == target) {
//             return l;
//         }
//         return -1;
//     }
// };

// 找左边界
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int mid, l = 0, r = nums.size() - 1;
        while (l < r) {
            mid = l + r >> 1;
            if (nums[mid] >= target) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        if (nums[l] == target) {
            return l;
        }
        return -1;
    }
};

File: codes/array/704.BinarySearch/BinarySearch.go
package main

import "fmt"

// 左闭右开版本
func search(nums []int, target int) int {
	
	l := 0
	r := len(nums)
	i := (l + r) / 2
	if target < nums[0] || target > nums[len(nums)-1] {
		return -1
	}
	for l < r {
		if nums[i] == target {
			return i
		} else if nums[i] < target {
			l = i + 1
		} else {
			r = i
		}
		i = (l + r) / 2
	}
	return -1
}

func main(){
	nums:=[]int{-1,0,3,5,9,12}
	fmt.Println(search(nums,9))
}